### PR TITLE
Closes #505 filter funnels by events as well as actions

### LINKED ISSUE #505: Filter funnels by events as well as actions
**Is your feature request related to a problem? Please describe.**
I don't want to create actions for each event to get a funnel.

**Describe the solution you'd like**
I'd like the exact same interface (components) for selecting Actions/Events in /trends in Funnels.

**Describe alternatives you've considered**
Creating actions.

**Additional context**
Split out from #444 as dates and events were slightly unrelated.

***Thank*** you for your feature request - we love each and every one!

### PR DESCRIPTION
**Pull request by timgl**
Draft for moving funnels to a 'filters' style mechanism for storing orders and such. This will allow us to create funnels with events as well as actions. It'll also make #546 and #477 easier.

@mariusandra Would love generic thoughts around the Kea stuff and the general approach to using filters to construct funnels.

Probably still todo before merging
- [ ] I think Filters should be a model as well, rather than just a dict we throw about to make sure we have a good structure and typing. Arguable whether to do in this PR 
- [x] Use ordering to sort steps in filter both in frontend and backend
- [ ] Move test_funnel.py tests to model tests file

### CODE REVIEW

File: frontend/src/scenes/trends/ActionFilter/ActionFilter.js [hunk 1]
```diff
@@ -3,9 +3,9 @@ import { useActions, useValues } from 'kea'
 import { entityFilterLogic } from './actionFilterLogic'
 import { ActionFilterRow } from './ActionFilterRow'
 
-export function ActionFilter(props) {
-    const { allFilters, filters } = useValues(entityFilterLogic)
-    const { createNewFilter, initializeLocalFilters } = useActions(entityFilterLogic)
+export function ActionFilter({ setFilters, defaultFilters, showMaths }) {
+    const { allFilters, filters } = useValues(entityFilterLogic({ setFilters, defaultFilters }))
+    const { createNewFilter, initializeLocalFilters } = useActions(entityFilterLogic({ setFilters, defaultFilters }))
```
Comment (mariusandra) [inline frontend/src/scenes/trends/ActionFilter/ActionFilter.js:8]: I haven't written about it in the docs, but seeing this I get a feeling that passing `setFilters`, which in the end is just a kea action (or sometimes `(filters) => setFunnel({ filters })`?), as `props` to the logic seems like there's a lot of entanglement happening here and it's hard to follow along when reading the code.

It could be perfectly fine though. The mental model I usually look for is this:
- if everything follows a *straight line* from react via an action to the logic to its reducers/selectors and then back to react via a value... it's good
- if there's *loop* in the line, either by using react as a conveyor belt to move actions around different logics or using react as something that reacts to value changes and then dispatches new actions (like the useEffect a few lines below)... it smells fishy and can probably be simplified

**Review by mariusandra — CHANGES_REQUESTED** [review]
Hey! Here's my feedback for the JS/Kea part. I didn't read through the python code, I trust you know what you're doing there :).

File: frontend/src/scenes/trends/ActionFilter/ActionFilter.js (hunk 1 above)
Comment (timgl) [reply frontend/src/scenes/trends/ActionFilter/ActionFilter.js:8]: Managed to do everything else but struggled with this one.

I guess at the moment it's more 2. The filters go from a funnel -> ActionFilter (where a user can add/remove actions/events) -> back to funnel to be stored/saved.

ActionFilter takes a filter dict as input, and returns a changed filter dict as output to the setFilter function. This does run via react, but because it's in 2 places (funnels and trends) I struggled to work out how to only do this via kea. Do you have any suggestions there?

**Review by mariusandra — APPROVED** [review]
I'd still try to refactor it a bit - to at least move the `initializeLocalFilters` in `useEffect` into the logic. I guess the part with `setFilters` via props is fine in the end...

File: frontend/src/scenes/trends/ActionFilter/ActionFilter.js (hunk 1 above)
Comment (EDsCODE) [reply frontend/src/scenes/trends/ActionFilter/ActionFilter.js:8]: I thought I'd chime in here as the actionfilter is giving some difficulty. 
1. The reason the action filters duplicates the original filters it's passed in is to preserve the ordering when users create the filters because we have multiple now. If we re-rendered them based on the dict it wouldn't always place filters as the last one. 
2. The useeffect was a big edge case that I didn't want to use but placed in there to solve the instance upon initial load where the parent trendlogic hadn't retrieved any filters yet but gave the actionfilter logic an empty filter array. 

Would definitely prefer not to use it, but wasn't fully sure how to set up the entitylogic to wait for the initial load of the filters to complete

Comment (timgl) [reply frontend/src/scenes/trends/ActionFilter/ActionFilter.js]: I think I've managed to make this work. Basically all `filters` is at the moment is just a selectors that takes `allFilters` and puts it into a dict that looks like `{actions: [], events: []}`. This massively simplified the Logic everywhere, and also allowed me (I think!) to get rid of useEffect.

**Comment by timgl** [comment]
@mariusandra Ready for another quick review. As this is getting pretty long i've created #572 to sort out the typing of the filters object properly.

**Comment by mariusandra** [comment]
Ah, great to see all the deleted listener, hooks, etc code in the last commits! 🤩👍 
It does look a lot neater now. Again, I didn't run and QA it, but code wise I have nothing more to add.

File: frontend/src/scenes/trends/ActionFilter/ActionFilter.js (hunk 1 above)
Comment (mariusandra) [reply frontend/src/scenes/trends/ActionFilter/ActionFilter.js]: @EDsCODE yeah, there's no good "wait until x is available" pattern that I know of, other than just using a listener to update the values once they do become available.

It *might* be worth looking into adding a new event, `onPropsChanged` or something like that which will get called every time the props are updated. However that might be more of a hack than a solution, I don't know :).

Tim's last commits did simplify a lot and hopefully removed the need for something like this entirely.

**Comment by timgl** [comment]
@Tannergoods This is ready for QA -- it's touched both trends and funnels, as they now use the same component.

You can create random events by doing `posthog.capture('cool event')` in /demo

**Comment by Tannergoods** [comment]
When navigating to Trends page does now load, it takes a refresh to get there.

You can navigate to /funnel but cannot navigate to /New-funnel at the moment

**Comment by timgl** [comment]
@Tannergoods This is ready for re-QA

**Comment by Tannergoods** [comment]
I don't think this error is related to this pr?

Posthog

Other than the above it works like a dream, very easy set up I also like the updated design to funnel building feels simpler

**Comment by timgl** [comment]
When does that error happen?

**Comment by Tannergoods** [comment]
It did happen when I saved a new funnel but I can't reproduce it now?

**Comment by timgl** [comment]
Hm okay, let's merge this b/c it's blocking other things and wait for it to re-appear
